File: tools/rp2040_hil.py
#!/usr/bin/env python3
"""Run the fixed Conduit RP2040 USB-CDC HIL exchange."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import pathlib
import select
import struct
import subprocess
import sys
import termios
import time


REQUEST = struct.Struct(">4sH16s32sI")
HEADER = struct.Struct(">4sH16s32s32s32s16sQBIH")
EVENT = struct.Struct(">4sH16s32s16sQIIBHBH16s")
PROTOCOL_VERSION = 1
EXPECTED_PRODUCT_PATH = pathlib.Path("/dev/serial/by-id")
ROOT = pathlib.Path(__file__).resolve().parents[1]
FIRMWARE_ROOT = ROOT / "firmware/conduit-rp2040-hil"
FIRMWARE_INPUTS = (
    "../../Cargo.lock",
    "../../Cargo.toml",
    "../../crates/conduit-core/Cargo.toml",
    "../../crates/conduit-core/src",
    "../../crates/conduit-embedded/Cargo.toml",
    "../../crates/conduit-embedded/src/lib.rs",
    "Cargo.toml",
    "build.rs",
    "memory.x",
    "src/lib.rs",
    "src/main.rs",
)

EVENT_KINDS = {
    1: "allocation-prepared",
    2: "node-prepared",
    3: "run-started",
    4: "decision",
    5: "value-accepted",
    6: "value-consumed",
    7: "pressure-entered",
    8: "pressure-cleared",
    9: "node-completed",
    10: "cancellation-requested",
    11: "run-succeeded",
    12: "run-cancelled",
    13: "run-failed",
}


def parser() -> argparse.ArgumentParser:
    value = argparse.ArgumentParser()
    value.add_argument("--port", default=os.environ.get("CONDUIT_RP2040_HIL_PORT"))
    value.add_argument("--expected-plan-hash")
    value.add_argument("--expected-firmware-identity")
    value.add_argument("--maximum-decisions", type=int, default=64)
    value.add_argument("--timeout-seconds", type=float, default=10.0)
    value.add_argument("--probe", action="store_true")
    value.add_argument("--require-hardware", action="store_true")
    return value


def discover_port() -> str | None:
    if not EXPECTED_PRODUCT_PATH.is_dir():
        return None
    matches = sorted(
        entry
        for entry in EXPECTED_PRODUCT_PATH.iterdir()
        if "Conduit_RP2040_HIL" in entry.name
        or "conduit-rp2040-hil" in entry.name.lower()
    )
    return str(matches[0]) if len(matches) == 1 else None


def configure_raw(fd: int) -> None:
    attributes = termios.tcgetattr(fd)
    attributes[0] = 0
    attributes[1] = 0
    attributes[2] = termios.CS8 | termios.CREAD | termios.CLOCAL
    attributes[3] = 0
    attributes[4] = termios.B115200
    attributes[5] = termios.B115200
    attributes[6][termios.VMIN] = 0
    attributes[6][termios.VTIME] = 0
    termios.tcsetattr(fd, termios.TCSANOW, attributes)
    termios.tcflush(fd, termios.TCIOFLUSH)


def read_exact(fd: int, length: int, deadline: float) -> bytes:
    output = bytearray()
    while len(output) < length:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise TimeoutError(f"timed out after {len(output)} of {length} bytes")
        readable, _, _ = select.select([fd], [], [], remaining)
        if readable:
            chunk = os.read(fd, length - len(output))
            if chunk:
                output.extend(chunk)
    return bytes(output)


def hash_bytes(value: str) -> bytes:
    normalized = value.removeprefix("sha256:")
    decoded = bytes.fromhex(normalized)
    if len(decoded) != 32:
        raise ValueError("expected plan hash must be 32 bytes")
    return decoded


def current_firmware_identity(
    target: str = "thumbv6m-none-eabi", profile: str = "release"
) -> bytes:
    digest = hashlib.sha256()
    for relative in FIRMWARE_INPUTS:
        source = (FIRMWARE_ROOT / relative).resolve()
        if source.is_dir():
            inputs = [
                (f"{relative}/{path.relative_to(source).as_posix()}", path)
                for path in sorted(source.rglob("*"))
                if path.is_file()
            ]
        else:
            inputs = [(relative, source)]
        for label, path in inputs:
            content = path.read_bytes()
            digest.update(label.encode("utf-8"))
            digest.update(b"\0")
            digest.update(len(content).to_bytes(8, "big"))
            digest.update(content)
    for label, content in (
        ("cargo-target", target.encode("utf-8")),
        ("cargo-profile", profile.encode("utf-8")),
        ("rustc-version", subprocess.check_output(["rustc", "-vV"])),
    ):
        digest.update(label.encode("utf-8"))
        digest.update(b"\0")
        digest.update(len(content).to_bytes(8, "big"))
        digest.update(content)
    return digest.digest()


def main() -> int:
    arguments = parser().parse_args()
    port = arguments.port or discover_port()
    if arguments.probe:
        report = {
            "schema": "conduit.rp2040-hil-probe/v1",
            "detected": port is not None,
            "port": port,
            "expected_firmware_identity": (
                f"sha256:{current_firmware_identity().hex()}"
            ),
        }
        print(json.dumps(report, sort_keys=True))
        return 0 if port is not None or not arguments.require_hardware else 2
    if not port:
        message = "no unique Conduit RP2040 HIL USB-CDC device detected"
        if arguments.require_hardware:
            print(message, file=sys.stderr)
            return 2
        print(json.dumps({"executed": False, "reason": message}, sort_keys=True))
        return 0
    if not arguments.expected_plan_hash:
        print("--expected-plan-hash is required for an HIL run", file=sys.stderr)
        return 2
    expected_plan = hash_bytes(arguments.expected_plan_hash)
    expected_firmware = (
        hash_bytes(arguments.expected_firmware_identity)
        if arguments.expected_firmware_identity
        else current_firmware_identity()
    )
    nonce = os.urandom(16)
    request = REQUEST.pack(
        b"CNH1",
        PROTOCOL_VERSION,
        nonce,
        expected_plan,
        arguments.maximum_decisions,
    )
    fd = os.open(port, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)
    try:
        configure_raw(fd)
        os.write(fd, request)
        deadline = time.monotonic() + arguments.timeout_seconds
        header = HEADER.unpack(read_exact(fd, HEADER.size, deadline))
        (
            magic,
            version,
            response_nonce,
            plan,
            firmware_identity,
            capability_report_hash,
            boot_id,
            run_sequence,
            status,
            decisions,
            count,
        ) = header
        if (
            magic != b"CNR1"
            or version != PROTOCOL_VERSION
            or response_nonce != nonce
            or plan != expected_plan
            or firmware_identity != expected_firmware
            or capability_report_hash == bytes(32)
            or status != 1
        ):
            raise RuntimeError("HIL response header failed identity or status validation")
        events = []
        for expected_sequence in range(count):
            fields = EVENT.unpack(read_exact(fd, EVENT.size, deadline))
            (
                event_magic,
                event_version,
                event_nonce,
                event_plan,
                event_boot,
                event_run,
                sequence,
                tick,
                subject_kind,
                subject_index,
                kind,
                value_length,
                value,
            ) = fields
            if (
                event_magic != b"CNE1"
                or event_version != PROTOCOL_VERSION
                or event_nonce != nonce
                or event_plan != expected_plan
                or event_boot != boot_id
                or event_run != run_sequence
                or sequence != expected_sequence
                or kind not in EVENT_KINDS
                or value_length > len(value)
            ):
                raise RuntimeError("HIL event attribution or sequence validation failed")
            events.append(
                {
                    "sequence": sequence,
                    "tick": tick,
                    "subject_kind": subject_kind,
                    "subject_index": subject_index,
                    "kind": EVENT_KINDS[kind],
                    "value": value[:value_length].hex(),
                }
            )
        kinds = {event["kind"] for event in events}
        required = {
            "allocation-prepared",
            "node-prepared",
            "run-started",
            "value-accepted",
            "value-consumed",
            "pressure-entered",
            "pressure-cleared",
            "node-completed",
            "run-succeeded",
        }
        if not required.issubset(kinds):
            raise RuntimeError(f"HIL evidence omitted {sorted(required - kinds)}")
        accepted = [
            bytes.fromhex(event["value"])
            for event in events
            if event["kind"] == "value-accepted"
        ]
        if accepted != [(42).to_bytes(4, "big"), b"\x01"]:
            raise RuntimeError("HIL semantic values differ from the representative oracle")
        report = {
            "schema": "conduit.rp2040-hil-report/v1",
            "executed": True,
            "port": port,
            "plan_hash": f"sha256:{plan.hex()}",
            "firmware_identity": f"sha256:{firmware_identity.hex()}",
            "capability_report_hash": f"sha256:{capability_report_hash.hex()}",
            "boot_id": boot_id.hex(),
            "run_sequence": run_sequence,
            "decisions": decisions,
            "evidence_records": count,
            "normalized": {
                "values": [value.hex() for value in accepted],
                "pressure_entered": "pressure-entered" in kinds,
                "pressure_cleared": "pressure-cleared" in kinds,
                "terminal": "run-succeeded",
            },
        }
        print(json.dumps(report, indent=2, sort_keys=True))
        return 0
    finally:
        os.close(fd)


if __name__ == "__main__":
    raise SystemExit(main())

File: xtask/src/commands/rp2040_hil.rs
use std::{fs, io::Read, path::Path};

use serde_json::json;

use crate::commands::embedded_gate::current_firmware_identity;

pub struct Rp2040HilOptions {
    pub port: Option<String>,
    pub expected_plan_hash: Option<String>,
    pub expected_firmware_identity: Option<String>,
    pub maximum_decisions: u32,
    pub timeout_seconds: f64,
    pub probe: bool,
    pub require_hardware: bool,
}

fn discover_port() -> Option<String> {
    let dev_dir = Path::new("/dev/serial/by-id");
    if !dev_dir.is_dir() {
        return None;
    }
    if let Ok(entries) = fs::read_dir(dev_dir) {
        let matches: Vec<_> = entries
            .flatten()
            .filter(|e| {
                let name = e.file_name().to_string_lossy().to_string();
                name.contains("Conduit_RP2040_HIL")
                    || name.to_lowercase().contains("conduit-rp2040-hil")
            })
            .collect();
        if matches.len() == 1 {
            return Some(matches[0].path().to_string_lossy().to_string());
        }
    }
    None
}

fn hash_bytes(val: &str) -> Result<Vec<u8>, String> {
    let normalized = val.strip_prefix("sha256:").unwrap_or(val);
    let decoded = hex::decode(normalized).map_err(|e| e.to_string())?;
    if decoded.len() != 32 {
        return Err("expected plan hash must be 32 bytes".to_string());
    }
    Ok(decoded)
}

#[cfg(unix)]
fn configure_raw_termios(fd: std::os::unix::io::RawFd) -> Result<(), String> {
    unsafe {
        let mut term: libc::termios = std::mem::zeroed();
        if libc::tcgetattr(fd, &mut term) != 0 {
            return Err("tcgetattr failed".to_string());
        }
        term.c_iflag = 0;
        term.c_oflag = 0;
        term.c_cflag = libc::CS8 | libc::CREAD | libc::CLOCAL;
        term.c_lflag = 0;
        libc::cfsetispeed(&mut term, libc::B115200);
        libc::cfsetospeed(&mut term, libc::B115200);
        term.c_cc[libc::VMIN] = 0;
        term.c_cc[libc::VTIME] = 0;
        if libc::tcsetattr(fd, libc::TCSANOW, &term) != 0 {
            return Err("tcsetattr failed".to_string());
        }
        libc::tcflush(fd, libc::TCIOFLUSH);
    }
    Ok(())
}

pub fn run(
    workspace_root: &Path,
    opts: Rp2040HilOptions,
) -> Result<(), Box<dyn std::error::Error>> {
    let port = opts.port.or_else(discover_port);

    if opts.probe {
        let firmware_id =
            current_firmware_identity(workspace_root, "thumbv6m-none-eabi", "release")?;
        let report = json!({
            "schema": "conduit.rp2040-hil-probe/v1",
            "detected": port.is_some(),
            "port": port,
            "expected_firmware_identity": format!("sha256:{}", hex::encode(firmware_id)),
        });
        println!("{}", serde_json::to_string(&report)?);
        if port.is_none() && opts.require_hardware {
            std::process::exit(2);
        }
        return Ok(());
    }

    let port_str = match port {
        Some(p) => p,
        None => {
            let message = "no unique Conduit RP2040 HIL USB-CDC device detected";
            if opts.require_hardware {
                eprintln!("{message}");
                std::process::exit(2);
            } else {
                let report = json!({ "executed": false, "reason": message });
                println!("{}", serde_json::to_string(&report)?);
                return Ok(());
            }
        }
    };

    let plan_str = opts
        .expected_plan_hash
        .ok_or("--expected-plan-hash is required for an HIL run")?;
    let expected_plan = hash_bytes(&plan_str)?;

    let expected_firmware = match opts.expected_firmware_identity {
        Some(ref f) => hash_bytes(f)?,
        None => current_firmware_identity(workspace_root, "thumbv6m-none-eabi", "release")?,
    };

    let mut nonce = [0u8; 16];
    getrandom_bytes(&mut nonce)?;

    // REQUEST struct: magic(4s), version(H), nonce(16s), plan(32s), max_decisions(I)
    let mut request = Vec::with_capacity(58);
    request.extend_from_slice(b"CNH1");
    request.extend_from_slice(&1u16.to_be_bytes()); // PROTOCOL_VERSION = 1
    request.extend_from_slice(&nonce);
    request.extend_from_slice(&expected_plan);
    request.extend_from_slice(&opts.maximum_decisions.to_be_bytes());

    #[cfg(unix)]
    {
        use std::os::unix::ffi::OsStrExt;
        let path_c = std::ffi::CString::new(Path::new(&port_str).as_os_str().as_bytes())?;
        let fd = unsafe {
            libc::open(
                path_c.as_ptr(),
                libc::O_RDWR | libc::O_NOCTTY | libc::O_NONBLOCK,
            )
        };
        if fd < 0 {
            return Err(format!("failed to open serial port {port_str}").into());
        }

        struct FdGuard(libc::c_int);
        impl Drop for FdGuard {
            fn drop(&mut self) {
                unsafe { libc::close(self.0) };
            }
        }
        let _guard = FdGuard(fd);

        configure_raw_termios(fd)?;

        let written = unsafe { libc::write(fd, request.as_ptr() as *const _, request.len()) };
        if written < 0 || written as usize != request.len() {
            return Err("failed to write complete CNH1 request to serial port".into());
        }

        let deadline =
            std::time::Instant::now() + std::time::Duration::from_secs_f64(opts.timeout_seconds);

        fn read_exact_fd(
            fd: libc::c_int,
            len: usize,
            deadline: std::time::Instant,
        ) -> Result<Vec<u8>, String> {
            let mut buf = vec![0u8; len];
            let mut read_bytes = 0;
            while read_bytes < len {
                let now = std::time::Instant::now();
                if now >= deadline {
                    return Err(format!("timed out after {read_bytes} of {len} bytes"));
                }
                let timeout = deadline - now;
                let mut pfd = libc::pollfd {
                    fd,
                    events: libc::POLLIN,
                    revents: 0,
                };
                let ret = unsafe { libc::poll(&mut pfd, 1, timeout.as_millis() as libc::c_int) };
                if ret > 0 && (pfd.revents & libc::POLLIN) != 0 {
                    let n = unsafe {
                        libc::read(
                            fd,
                            buf[read_bytes..].as_mut_ptr() as *mut _,
                            len - read_bytes,
                        )
                    };
                    if n > 0 {
                        read_bytes += n as usize;
                    }
                }
            }
            Ok(buf)
        }

        // Header format (179 bytes):
        // magic(4s), version(u16), response_nonce(16s), plan(32s), firmware(32s), capability(32s), boot_id(16s), run_sequence(u64), status(u8), decisions(u32), count(u16)
        let header_buf = read_exact_fd(fd, 179, deadline)?;
        let magic = &header_buf[0..4];
        let version = u16::from_be_bytes(header_buf[4..6].try_into().unwrap());
        let response_nonce = &header_buf[6..22];
        let plan = &header_buf[22..54];
        let firmware_identity = &header_buf[54..86];
        let capability_report_hash = &header_buf[86..118];
        let boot_id = &header_buf[118..134];
        let run_sequence = u64::from_be_bytes(header_buf[134..142].try_into().unwrap());
        let status = header_buf[142];
        let decisions = u32::from_be_bytes(header_buf[143..147].try_into().unwrap());
        let count = u16::from_be_bytes(header_buf[147..149].try_into().unwrap());

        if magic != b"CNR1"
            || version != 1
            || response_nonce != nonce
            || plan != expected_plan
            || firmware_identity != expected_firmware
            || capability_report_hash == [0u8; 32]
            || status != 1
        {
            return Err("HIL response header failed identity or status validation".into());
        }

        let event_kinds = [
            (1, "allocation-prepared"),
            (2, "node-prepared"),
            (3, "run-started"),
            (4, "decision"),
            (5, "value-accepted"),
            (6, "value-consumed"),
            (7, "pressure-entered"),
            (8, "pressure-cleared"),
            (9, "node-completed"),
            (10, "cancellation-requested"),
            (11, "run-succeeded"),
            (12, "run-cancelled"),
            (13, "run-failed"),
        ];
        let event_kinds_map: std::collections::BTreeMap<u8, &str> =
            event_kinds.into_iter().collect();

        let mut events = Vec::new();
        for expected_sequence in 0..(count as u64) {
            // Event struct format (108 bytes):
            // magic(4s), version(u16), nonce(16s), plan(32s), boot(16s), run(u64), sequence(u32), tick(u32), subject_kind(u8), subject_index(u16), kind(u8), value_length(u16), value(16s)
            let ev_buf = read_exact_fd(fd, 108, deadline)?;
            let ev_magic = &ev_buf[0..4];
            let ev_version = u16::from_be_bytes(ev_buf[4..6].try_into().unwrap());
            let ev_nonce = &ev_buf[6..22];
            let ev_plan = &ev_buf[22..54];
            let ev_boot = &ev_buf[54..70];
            let ev_run = u64::from_be_bytes(ev_buf[70..78].try_into().unwrap());
            let sequence = u32::from_be_bytes(ev_buf[78..82].try_into().unwrap());
            let tick = u32::from_be_bytes(ev_buf[82..86].try_into().unwrap());
            let subject_kind = ev_buf[86];
            let subject_index = u16::from_be_bytes(ev_buf[87..89].try_into().unwrap());
            let kind = ev_buf[89];
            let value_length = u16::from_be_bytes(ev_buf[90..92].try_into().unwrap()) as usize;
            let val_bytes = &ev_buf[92..108];

            if ev_magic != b"CNE1"
                || ev_version != 1
                || ev_nonce != nonce
                || ev_plan != expected_plan
                || ev_boot != boot_id
                || ev_run != run_sequence
                || (sequence as u64) != expected_sequence
                || !event_kinds_map.contains_key(&kind)
                || value_length > val_bytes.len()
            {
                return Err("HIL event attribution or sequence validation failed".into());
            }

            events.push(json!({
                "sequence": sequence,
                "tick": tick,
                "subject_kind": subject_kind,
                "subject_index": subject_index,
                "kind": event_kinds_map[&kind],
                "value": hex::encode(&val_bytes[..value_length]),
            }));
        }

        let received_kinds: std::collections::BTreeSet<&str> = events
            .iter()
            .filter_map(|e| e.get("kind").and_then(serde_json::Value::as_str))
            .collect();

        let required = [
            "allocation-prepared",
            "node-prepared",
            "run-started",
            "value-accepted",
            "value-consumed",
            "pressure-entered",
            "pressure-cleared",
            "node-completed",
            "run-succeeded",
        ];

        for req in required {
            if !received_kinds.contains(req) {
                return Err(format!("HIL evidence omitted missing required event {req}").into());
            }
        }

        let accepted: Vec<Vec<u8>> = events
            .iter()
            .filter(|e| e.get("kind").and_then(serde_json::Value::as_str) == Some("value-accepted"))
            .filter_map(|e| e.get("value").and_then(serde_json::Value::as_str))
            .filter_map(|v| hex::decode(v).ok())
            .collect();

        let expected_accepted = vec![42u32.to_be_bytes().to_vec(), vec![0x01]];
        if accepted != expected_accepted {
            return Err("HIL semantic values differ from the representative oracle".into());
        }

        let report = json!({
            "schema": "conduit.rp2040-hil-report/v1",
            "executed": true,
            "port": port_str,
            "plan_hash": format!("sha256:{}", hex::encode(plan)),
            "firmware_identity": format!("sha256:{}", hex::encode(firmware_identity)),
            "capability_report_hash": format!("sha256:{}", hex::encode(capability_report_hash)),
            "boot_id": hex::encode(boot_id),
            "run_sequence": run_sequence,
            "decisions": decisions,
            "evidence_records": count,
            "normalized": {
                "values": accepted.iter().map(hex::encode).collect::<Vec<_>>(),
                "pressure_entered": received_kinds.contains("pressure-entered"),
                "pressure_cleared": received_kinds.contains("pressure-cleared"),
                "terminal": "run-succeeded",
            },
        });

        println!("{}", serde_json::to_string_pretty(&report)?);
    }

    Ok(())
}

fn getrandom_bytes(buf: &mut [u8]) -> Result<(), String> {
    #[cfg(unix)]
    {
        use std::fs::File;
        let mut f = File::open("/dev/urandom").map_err(|e| e.to_string())?;
        f.read_exact(buf).map_err(|e| e.to_string())?;
        Ok(())
    }
    #[cfg(not(unix))]
    {
        Err("platform random not available".to_string())
    }
}
